drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,373 @@
+	.workmode_mask = 0x02u,
+	.workmode_pwm_val = 0x02u,
+	.ranges = dcdc3_ranges,
+	.num_ranges = ARRAY_SIZE(dcdc3_ranges),
+};
+
+static const struct linear_range ldo2_ranges[] = {
+	LINEAR_RANGE_INIT(1800000u, 100000u, 0x00u, 0x0Fu),
+};
+
+static const struct regulator_axp192_desc ldo2_desc = {
+	.enable_reg = AXP192_REG_DCDC123_LDO23_CONTROL,
+	.enable_mask = 0x04u,
+	.enable_val = 0x04u,
+	.vsel_reg = AXP192_REG_LDO23_VOLTAGE,
+	.vsel_mask = 0xF0u,
+	.vsel_bitpos = 4u,
+	.max_ua = 200u,
+	.workmode_reg = 0u,
+	.workmode_mask = 0u,
+	.ranges = ldo2_ranges,
+	.num_ranges = ARRAY_SIZE(ldo2_ranges),
+};
+
+static const struct linear_range ldo3_ranges[] = {
+	LINEAR_RANGE_INIT(1800000u, 100000u, 0x00u, 0x0Fu),
+};
+
+static const struct regulator_axp192_desc ldo3_desc = {
+	.enable_reg = AXP192_REG_DCDC123_LDO23_CONTROL,
+	.enable_mask = 0x08u,
+	.enable_val = 0x08u,
+	.vsel_reg = AXP192_REG_LDO23_VOLTAGE,
+	.vsel_mask = 0x0Fu,
+	.vsel_bitpos = 0u,
+	.max_ua = 200u,
+	.workmode_reg = 0u,
+	.workmode_mask = 0u,
+	.ranges = ldo3_ranges,
+	.num_ranges = ARRAY_SIZE(ldo3_ranges),
+};
+
+int axp192_enable(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->mfd->config;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "Enabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0x%02x mask=0x%02x", (uint32_t)config->desc->enable_reg,
+		     (uint32_t)config->desc->enable_val, (uint32_t)config->desc->enable_mask);
+
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, config->desc->enable_val);
+
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to enable regulator");
+	}
+
+	return ret;
+}
+
+int axp192_disable(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->mfd->config;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "Disabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0 mask=0x%x", (uint32_t)config->desc->enable_reg,
+		     (uint32_t)config->desc->enable_mask);
+
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, 0u);
+
+	if (ret != 0) {
+		LOG_INST_ERR(config->log, "Failed to disable regulator");
+	}
+
+	return ret;
+}
+
+unsigned int axp192_count_voltages(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_values_count(config->desc->ranges, config->desc->num_ranges);
+}
+
+int axp192_list_voltage(const struct device *dev, unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, idx,
+					    volt_uv);
+}
+
+int axp192_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
```

Comment:
these all need to be static

Reply:
You have really found a lot of fundamental unattractiveness. Thank you so much for putting so much effort into this!

Reply:
Fixed with latest commit